pandas/io/parsers/base_parser.py

```diff
@@ -855,8 +856,10 @@ def _check_data_length(
         data: list of array-likes containing the data column-wise.
         """
         if not self.index_col and len(columns) != len(data) and columns:
+            # without this cast mypy thinks that last_entry could still be an ndarray
+            last_entry = cast(Scalar, data[-1])
             if len(columns) == len(data) - 1 and np.all(
-                (is_object_dtype(data[-1]) and data[-1] == "") | isna(data[-1])
+                (is_object_dtype(last_entry) and last_entry == "") | isna(last_entry)
```

Comment:
if last_entry is Scalar then `is_object_dtype` doesn't make much sense here

Reply:
Okay, I will just add a mypy ignore comment. The issue seems that the left part of the `|` could be an `ExtensionArray` which doesn't match any numpy-overload for `__ror__` (isna can return an np.ndframe)

Reply:
i actually meant that the code might be wrong/simplifiable.  i.e. this might be a case where typing actually catches something useful

Reply:
I'll look into it! I think strictly speaking, it could be used: if `data` is a 2D ndarray, `data[-1]` is still a ndarray.

Reply:
`is_object_dtype` seems to handle scalars:
https://github.com/pandas-dev/pandas/blob/2d7f86a507a0dcc496e74b38ca3013e8e8d28d7b/pandas/core/dtypes/common.py#L1623

Reply:
The tests that hit this line have always data `list[np.ndarray]`, meaning data[-1] is never a `Scalar` (at least in the tests).